### ISSUE TITLE
Make par less intrusive.

### ISSUE THREAD

**Issue by JackDandy** [nzbget/nzbget#290]
Ref: **NZBGet 18.0-testing**

Hello,

Scenario:
1) A download completes and requires par, that processing begins.
2) Other downloads complete while (1) is active, these also requires par, and are set [pp-queued]
3) Another download completes and does not require par, regardless, that item is set [pp-queued]

Can the download (3) that completes and does not need par, be instead run in a separate thread so that it can complete pp and not get set queued? I guess this is doable given that when an item needs a parset, a further download is made for those files, therefore, it is known whether or not a parset is required.

Further: There is no way to pause a pp item, **even** if it has not even begun a par process and is only queued - this is issue 1.

Issue 2: A par can take a great many hours to complete, therefore, all other downloads are stacked waiting for it to complete. Worse, if you have 10 downloads needing par - game over.

Thanks

**Comment by hugbug** [comment]
Everything is doable, that's the question of the efforts. Currently post-processor can work out one item at a time. Par-check is one of post-processing stages like many others. To post-process multiple items simultaneously the concept must be redesigned. We can try to do that if you don't abandon the discussion like in the other thread, smile.

Generally post-processing is mostly about disk operations. Performing multiple operations at once usually doesn't improve total time and may even worsen it. That's why the current design, and also because it's simpler of cause.

Why do you think it's better to do simultaneous repair and unpack of other item? Are you concerned about total processing time of all items or is it your use case merely a wish to get one specific download as soon as possible even if the post-processing queue is filled with other items? There could be different approaches in fullfilling these two requirements.

What for would you want to pause a pp-item which didn't start processing? There is a pause for post-processing in general and the pp-items can be reordered, their priorities are also taken into account when the next pp-item is picked for processing. Why pause then?

What you can do however is to cancel post-processing for an item, being it an active item or pp-queued (click on the item -> Actions -> Cancel post-processing). It will go into history where you can "post-process again" later.

I don't see a problem in issue 2. You have 10 items which will be processed some day, hopefully. Do you care about the processing order? You can change it and you can use priorities as well.

Or do you talk again about the case with "one specific download as soon as possible"? If that download is important you can give it a higher priority and it will be processed accordingly, **after the current pp-item is done**. I understand that last restriction may be too big if current pp-item requires hours of repairing. Hence the next paragraph.

If your main concern is indeed the **get one specific download as soon as possible**, then I can think of relatevely simple although probably not so elegant solution: if an item with force-priority enters post-processing queue the current item is automatically canceled (without moving to history) and the post-processor will pick up a next item which will be the force-priority item. Once the force-item is processed the previously active pp-item will be picked up next and the last active not finished post-processing stage will start over (for example par-repair or unpack, etc.). It will not be continued but rather restarted, that's the "not so elegant" part.

**Comment by JackDandy** [comment]
Um, I didn't "abandon" the other issue, I added what info I could, I didn't see a need to involve myself further as the issue had been received and entered dev discussion. As it's your project, you guys are intimate with it, as a user, I simply offer feedback on how the user experience is.

As for this thread... it's as simple as I said, however, I read it that there is too much effort in redesign therefore, not much can be done.

Just to be clear, 

1) It would be great if any number of downloads that do not need par are automatically processed on a separate thread to the one that is running par. This would **improve the throughput** processing downloads.

2) If something is pp-queued, manually pausing those items may be desired by the user. Right now, the user is told they cannot pause a pp item, why not? an external subprocess hasn't started - let me pause you.

Again, thanks.

**Comment by hugbug** [comment]
> This would improve the throughput processing downloads.

So that's your main concern, good to know that.
That claim has to be proofed actually. Par-repair uses all cpu cores, I'm not sure if doing something in parallel would improve the total processing time of all items. I guess the only way of verifying that is to implement your request. Let's do this.
Are you willing to perform the testing in described scenarios, with time tracking?

The reason of not being able to pause pp-item is a technical limitation of current implementation. For historical reason nzbs do not have pause state. Instead files contained within nzbs can be individually paused and resumed. When all files are paused the nzb is shown as paused. Since during post-processing there may be no files within nzb anymore there is no way to set it to pause state.
The internals of course can be changed to give nzb it's own pause flag but it has to be checked how that will play with pause states of individual files. Since no one asked about pausing pp-items before I never bothered to change current system. Maybe it's time.

**Comment by JackDandy** [comment]
Yeah, naturally, timing tests would be needed to prove if the concept actually works - it may not, and sure I can test here, I do need you to add timers and any other counters to the start and end of processes.

About pausing; I was thinking on a more crude level, let's say (excuse me if I am slightly off the technicals here) each step is discrete, that is, the downloading flow, the running of a script flow, the unpacking flow, the pp'ing flow, then a pause at its crudest level would be... check if flagged to pause in between each flow step, and if yes, don't proceed to the next flow - therefore, paused. However, if you can pause in a more granular way, then that would be far more responsive - and tbh useful - however, how you would get external par to pause would probably be another level. Hm, this is an idea, given a stdio can be suspended with keypress ctrl-s. ctrl-q could those be somehow sent to a running par that outputs a progress meter - nah, I am talking fantasy, aren't I

**Comment by hugbug** [comment]
There is command "pause post-processing" (in menu near main round button) which works exactly like this. Even processing of par-stages works.

NZBGet doesn't use external par, par-module is integrated directly into nzbget binary, it has the same GPL license and is written in C++ like nzbget, the integration has happened in v14.

However pausing of individual items is a more complicated thing. To par-process another item either another instance of par-check must be created or the main (which is currently the only one) instance must be stopped, which means the resuming will not work, so we practically talking about canceling of par-stage. Par-repair consumes memory and having multiple active instances (even paused) may be dangerous.

The pausing of pp-items, if implemented, will work on stage borders as you describe.

Which one of the two features ("par + other pp-processing" or "pausing pp-items") you prefer to have first?

**Comment by JackDandy** [comment]
Better (potentially. hopefully) throughput all day long, so ... "par + other pp-processing" please.

I hadn't noticed that little drop arrow, maybe a pause symbol ( II ) image above may make it stand out more.

**Comment by hugbug** [comment]
Tracking progress in #291 (Multi post-processing: non-exclusive par-repair).

Except for one little todo everything seems to work.

It's now technically possible to execute multiple post-processing jobs simultaneously. During internal testing I've configured the code to allow up to three jobs regardless of job kind. It was nice to see all jobs running in parallel doing par-check/repair, then unpacking and then executing post-processing scripts, all at the same time (I've paused post-processing, downloaded three items, damaged them manually by deleting few rar-files, then resumed the post-processing - all that to make the post-processing simultaneous).

I'm however not sure how practical that is and if this possibility should be opened to users. The current state of #291 allows only for max two jobs, one of each must be par-job and another non-par-job. It's not possible to have two par-jobs or two non-par-jobs unlike to what I was doing in my internal tests.

I'm now thinking whether I should replace that one new option **ParExclusive** with a set of other new options which would allow more flexible configuring. I'm certain it will be more difficult to explain how the options correlate.
- option **MaxPostJobs** - sets how many post-processing jobs (of any kind) can be executed simultaneously;
- option **MaxParJobs** - number of simultaneous par-jobs;
- option **MaxOtherJobs** - number of simultaneous non-par-jobs.

All three options are needed to cover the following example use cases:
1. **One post-processing job at a time**. Par-job blocks all others. This is current standard behavior. MaxPostJobs=1, MaxParJobs=1, MaxOtherJobs=1.
2. **One post-processing job at a time, not counting one par-job**. Par-job doesn't block other jobs. MaxPostJobs=2, MaxParJobs=1, MaxOtherJobs=1.
3. **Simultaneous processing of three jobs, with max two par-jobs**. One post-processing slot is reserved for non par-jobs; up to two par-jobs is possible. MaxPostJobs=3, MaxParJobs=2, MaxOtherJobs=3.

Can all this be explained to users or way too complex?

**Comment by JackDandy** [comment]
Sounds like the three options covers all bases, and will really benefit multi-core systems, I would be using (2) - a system with 8 cores would be able to give more.

Perhaps place the three new config items in a _sub_ advanced section underneath the new **ParExclusive** option, after all it is a nice simple set of default values for the other three settings. Advanced users can additionally power up - normal users can stick to the default or enable **ParExclusive**.

Maybe determine at run-time if a change to the config would be something the user would benefit from and then give them up to say 3 hints/warnings of this fact.  

I think the way you worded it above is a clear but maybe introduce why these options exist, see below...

---

A par job prevents other post processing tasks from running even if those tasks don't required par. The following settings may improve the processing of multiple releases if your system can run multiple tasks at the same time.

**Comment by hugbug** [comment]
The settings page in nzbget is too primitive to achieve this. Each option is shown independently and there are no conditions to show/hide other options. So it's either **ParExclusive** or the three others. Sorry.

**Comment by hugbug** [comment]
Alternatively we can offer only one option with radio-values to cover usual cases. 

The option could have name **PostProcessStrategy** with following values (only example possible names):
- Sequential
- OneParOneOther
- OneParTwoOthers
- TwoParTwoOthers
- ParallelThree
- ParallelFour

Followed by explanation what every value means.

**Comment by hugbug** [comment]
Or may be fewer possible values would make it easier to choose from:
- **Sequential** - current standard;
- **Balanced** - max two jobs, one par job, one other jobs;
- **Aggressive** - up to three simultaneous jobs, one par-job;
- **Rocket** - up to six simultaneous jobs, up to two par-jobs.

Because par-module can use multiple cores simultaneously it really makes little sense to allow multiple par-jobs.

**Comment by hugbug** [comment]
Will look like this:
![screen shot 2016-10-12 at 20 24 10](https://cloud.githubusercontent.com/assets/3368402/19322389/1fb1e972-90b9-11e6-89cb-66c94244f413.png)

**Comment by JackDandy** [comment]
Looks good...

Perhaps shorten that intro to make it easier to appreciate (and reduce tl;dr syndrome)...

"Sequential - downloaded items are post processed from a queue, one item at a time, to dedicate the most computer resources to each item. Therefore, a post process par repair will prevent other tasks from running even if those tasks don't required a par repair"

**Comment by JackDandy** [comment]
Looks good... perhaps shorten that intro to make it easier to appreciate (and reduce tl;dr syndrome)...
- **Sequential (_default_)** - downloaded items are post processed from a queue, one item at a time, to dedicate the most computer resources to each item. Therefore, a post process par repair will prevent another task from running even if the item does not require a par repair  
- **Balanced** - items that do not need par repair are post processed one at a time while par repair tasks also run simultaneously one after another. This means that a post process par repair will not prevent another task from running, but at a cost of using more computer resource (see footer note)  
- **Aggressive** - will simultaneously post process up to three items plus one par repair task
- **Rocket** -  will simultaneously post process up to six items plus one or two par repair tasks  

**Note**: computer resources are in heavy demand when unpacking with simultaneous tasks - double check the h/w is capable

---

Hope that helps

**Comment by hugbug** [comment]
Thanks for copywriting work; English isn't my best strength. Would you like to rewrite all other options? :wink: 

I'm planning to make Balanced default, except for very low powered devices.

In Aggressive and Rocket I meant **including** instead of **plus** about par tasks.

**Comment by hugbug** [comment]
Available for testing - [release 18.0-testing-r1820](https://github.com/nzbget/nzbget/releases/tag/v18.0-r1820).

**Comment by JackDandy** [comment]
Ha, inspired in the moment, I spent over an hour re-tuning that option text over and over - it would take a long time to do the same for all options. Yeah, balanced is a good for default. Installing r1820.

That functional testing system looks like a useful addition ;)

**Comment by JackDandy** [comment]
Can you make it so that if we are told that (1/10 selected) items cannot be unpaused because they are in PP state, then still unpause the remaining (9) selected items after the notification has been acknowledged.

(The above numbers serve to illustrate the point - don't put any thought into them other than that)

**Comment by hugbug** [comment]
> Can you make it so that if we are told that (1/10 selected) items cannot be unpaused because they are in PP state, then still unpause the remaining (9) selected items after the notification has been acknowledged.

Yes.

BTW any feedback to the post-processing strategies in r1820?

**Comment by JackDandy** [comment]
I don't see an issue with r1820 so far, I use the _Balanced_ strategy (like most users probably will)
